### ISSUE TITLE
Add a fork/major version for GHC >= 8.6

### ISSUE THREAD

**Issue by treeowl** [Gabriel439/Haskell-MMorph-Library#50]
As you've already seen, GHC 8.6 wins

```haskell
instance (MFunctor f, MFunctor g, forall m. Monad m => Monad (g m))
    => MFunctor (ComposeT f g)
```

But it has more to offer. We can change the `MonadTrans` instance to the less-constrained, and sometimes more efficient,

```haskell
instance (MonadTrans f, MonadTrans g, forall m. Monad m => Monad (g m)) => MonadTrans (ComposeT f g)
  where
    lift = ComposeT . lift . lift
```

and then even change the type of `hoist` to

```haskell
    hoist :: (Monad m, Monad n) => (forall a . m a -> n a) -> t m b -> t n b
```

which allows some important instances including `streamly` streams and Church-style free monad transformers.

Basically, aside from the fact that `QuantifiedConstraints` is a pain and a half for instance resolution, everything gets much nicer.

**Comment by Gabriel439** [comment]
Yes, I would like to take advantage of `QuantifiedConstraints` in some way.  It's a much nicer solution to the problem

**Comment by Gabriel439** [comment]
So I was thinking about ways to do this non-disruptively, and one possible solution is to just add a new `newtype` (e.g. `ComposeT2` for lack of a better name) with the new instance and then users can also use `deriving via (ComposeT2 f g)` to get the desired `MonadTrans` instance

**Comment by treeowl** [comment]
The problem is we really want to change the type of hoist, and that's
inherently disruptive.

**Comment by turion** [comment]
But the change in `hoist`s type is backwards compatible, because instances of the old version will still be instances of the new version, right?

**Comment by Gabriel439** [comment]
In general, any change to the type of `hoist` will be backwards-incompatible, either for people who author instances or for people who consume instances.

In this specific case, the proposed change to the type of `hoist` would be backwards-incompatible for consumers because we're adding a constraint

**Comment by turion** [comment]
> In this specific case, the proposed change to the type of `hoist` would be backwards-incompatible for consumers because we're adding a constraint

Makes sense. Maybe a transition scheme with a cabal flag `new-hoist` makes sense:

* Current version: Keep `hoist :: Monad m => ...`
* Version +1:
  * Introduce cabal flag `new-hoist` with default `false`
  * When flag is disabled `hoist :: Monad m => ...`, when it's enabled, `hoist :: (Monad m, Monad n) => ...`, with a haddock note that people are encouraged to switch the flag on
* Version +2:
  * Make flag default `true`
  * Add a deprecation note to `hoist :: Monad m => ...`
* Version +3: Remove flag and old version

This is extra effort during versions +1 and +2, but removes the need for a fork and prevents the issue of being stuck eternally.

**Comment by Gabriel439** [comment]
I'm personally still reluctant to change the type of `hoist` because it's still not clear that we need the extra `Monad` constraint

**Comment by turion** [comment]
> ```haskell
>     hoist :: (Monad m, Monad n) => (forall a . m a -> n a) -> t m b -> t n b
> ```
> 
> which allows some important instances including `streamly` streams and Church-style free monad transformers.

That I don't understand. Wouldn't you need `MInvariant` for this (https://github.com/Gabriel439/Haskell-MMorph-Library/pull/1).

**Comment by turion** [comment]
> I'm personally still reluctant to change the type of `hoist` because it's still not clear that we need the extra `Monad` constraint

Ah, right. The transition scheme could still be applied to changing the `ComposeT` type signature.

**Comment by treeowl** [comment]
@turion, no, I take a look in the `free` package.

**Comment by turion** [comment]
> @turion, no, I take a look in the `free` package.

Ah, yes, for Church-style free monads it seems like `Monad n` is needed. But for `streamly` I believe even `Monad n` is not enough, and you'd need `MInvariant` or so.

**Comment by Ericson2314** [comment]
`MonadTrans` did end up getting the quantified constraint, so I hope this package can get at least one too!
